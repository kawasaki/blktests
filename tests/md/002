#!/bin/bash
# SPDX-License-Identifier: GPL-3.0+
# Copyright (C) 2025 Oracle and/or its affiliates
#
# Test SCSI Atomic Writes with MD devices

. tests/scsi/rc
. common/scsi_debug
. common/xfs

DESCRIPTION="test md atomic writes"
QUICK=1

requires() {
	_have_driver scsi_debug
}

test() {
	local scsi_debug_params=(
		delay=0
		atomic_wr=1
		num_tgts=1
		add_host=4
		per_host_store=true
		dev_size_mb=16
	)

	echo "Running md_atomics_test"

	if ! _configure_scsi_debug "${scsi_debug_params[@]}"; then
		return 1
	fi

	scsi_0="${SCSI_DEBUG_DEVICES[0]}"
	scsi_1="${SCSI_DEBUG_DEVICES[1]}"
	scsi_2="${SCSI_DEBUG_DEVICES[2]}"
	scsi_3="${SCSI_DEBUG_DEVICES[3]}"

	_md_atomics_test "${SCSI_DEBUG_DEVICES[0]}" "${SCSI_DEBUG_DEVICES[1]}" \
			"${SCSI_DEBUG_DEVICES[2]}" "${SCSI_DEBUG_DEVICES[3]}"

	_exit_scsi_debug

	echo "Test complete"
}
