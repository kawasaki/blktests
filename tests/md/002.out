Running md/002
TEST 1 raid0 step 1 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid0 step 1 - Verify sysfs atomic attributes - pass
TEST 3 raid0 step 1 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid0 step 1 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid0 step 1 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid0 step 1 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid0 step 1 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid0 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid0 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid0 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid0 step 1 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid0 step 2 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid0 step 2 - Verify sysfs atomic attributes - pass
TEST 3 raid0 step 2 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid0 step 2 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid0 step 2 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid0 step 2 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid0 step 2 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid0 step 2 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid0 step 2 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid0 step 2 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid0 step 2 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid0 step 3 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid0 step 3 - Verify sysfs atomic attributes - pass
TEST 3 raid0 step 3 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid0 step 3 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid0 step 3 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid0 step 3 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid0 step 3 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid0 step 3 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid0 step 3 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid0 step 3 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid0 step 3 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid0 step 4 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid0 step 4 - Verify sysfs atomic attributes - pass
TEST 3 raid0 step 4 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid0 step 4 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid0 step 4 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid0 step 4 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid0 step 4 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid0 step 4 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid0 step 4 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid0 step 4 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid0 step 4 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid1 step 1 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid1 step 1 - Verify sysfs atomic attributes - pass
TEST 3 raid1 step 1 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid1 step 1 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid1 step 1 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid1 step 1 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid1 step 1 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid1 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid1 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid1 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid1 step 1 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid10 step 1 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid10 step 1 - Verify sysfs atomic attributes - pass
TEST 3 raid10 step 1 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid10 step 1 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid10 step 1 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid10 step 1 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid10 step 1 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid10 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid10 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid10 step 1 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid10 step 1 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid10 step 2 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid10 step 2 - Verify sysfs atomic attributes - pass
TEST 3 raid10 step 2 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid10 step 2 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid10 step 2 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid10 step 2 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid10 step 2 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid10 step 2 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid10 step 2 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid10 step 2 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid10 step 2 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid10 step 3 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid10 step 3 - Verify sysfs atomic attributes - pass
TEST 3 raid10 step 3 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid10 step 3 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid10 step 3 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid10 step 3 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid10 step 3 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid10 step 3 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid10 step 3 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid10 step 3 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid10 step 3 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
TEST 1 raid10 step 4 - Verify md sysfs atomic attributes matches - pass
TEST 2 raid10 step 4 - Verify sysfs atomic attributes - pass
TEST 3 raid10 step 4 - Verify md sysfs_atomic_write_max is equal to expected_atomic_write_max - pass
TEST 4 raid10 step 4 - Verify sysfs atomic_write_unit_max_bytes =  expected_atomic_write_unit_max - pass
TEST 5 raid10 step 4 - Verify sysfs atomic_write_unit_boundary_bytes = expected atomic_write_unit_boundary_bytes - pass
TEST 6 raid10 step 4 - Verify statx stx_atomic_write_unit_min - pass
TEST 7 raid10 step 4 - Verify statx stx_atomic_write_unit_max - pass
TEST 8 raid10 step 4 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 9 raid10 step 4 - perform a pwritev2 with size of sysfs_atomic_unit_max_bytes + LBS bytes with RWF_ATOMIC flag - pwritev2 should not be succesful - pass
TEST 10 raid10 step 4 - perform a pwritev2 with size of sysfs_atomic_unit_min_bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
pwrite: Invalid argument
TEST 11 raid10 step 4 - perform a pwritev2 with a size of sysfs_atomic_write_unit_max_bytes - LBS bytes with RWF_ATOMIC flag - pwritev2 should fail - pass
Test complete
